cmd/errtrace/main.go

```diff
@@ -173,27 +183,152 @@ func (cmd *mainCmd) Run(args []string) (exitCode int) {
+	if wd, err := os.Getwd(); err == nil {
+		workDir = wd + string(filepath.Separator)
+	}
+
+	for _, file := range files {
+		display := file
+		if workDir != "" {
+			// Not using filepath.Rel
+			// because we don't want any ".."s in the path.
+			display = strings.TrimPrefix(file, workDir)
+		}
+		if display == "-" {
+			display = "stdin"
+		}
+
 		req := fileRequest{
 			Format:   p.shouldFormat(),
 			Write:    p.Write,
 			List:     p.List,
-			Filename: file,
+			Filename: display,
+			Filepath: file,
 		}
 		if err := cmd.processFile(req); err != nil {
-			cmd.log.Printf("%s:%s", file, err)
+			cmd.log.Printf("%s:%s", display, err)
 			exitCode = 1
 		}
 	}
 
 	return exitCode
 }
 
+// expandPatterns turns the given list of patterns and files
+// into a list of paths to files.
+//
+// Arguments that are already files are returned as-is.
+// Arguments that are patterns are expanded using 'go list'.
+// As a special case for stdin, "-" is returned as-is.
+func expandPatterns(args []string) ([]string, error) {
+	var files, patterns []string
+	for _, arg := range args {
+		if arg == "-" {
+			files = append(files, arg)
+			continue
+		}
+
+		if info, err := os.Stat(arg); err == nil && !info.IsDir() {
+			files = append(files, arg)
+			continue
+		}
+
+		patterns = append(patterns, arg)
+	}
+
+	if len(patterns) > 0 {
+		pkgFiles, err := goListFiles(patterns)
+		if err != nil {
+			return nil, fmt.Errorf("go list: %w", err)
+		}
+
+		files = append(files, pkgFiles...)
+	}
+
+	return files, nil
+}
+
+var _execCommand = exec.Command
+
+func goListFiles(patterns []string) (files []string, err error) {
+	// If we use "go list" to find the files,
+	// it'll only report files that match the default build constraints.
+	//
+	// We want all files so we just request the package directories,
+	// and then walk them ourselves to find all the files.
+	//
+	// The -e flag makes 'go list' include erroneous packages,
+	// e.g. those with all files excluded by build constraints.
+	args := []string{"list", "-find", "-e", "-json"}
+	args = append(args, patterns...)
+
+	var stderr bytes.Buffer
+	cmd := _execCommand("go", args...)
+	cmd.Stderr = &stderr
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("create stdout pipe: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("start command: %w", err)
+	}
+
+	type packageInfo struct {
+		Dir               string
+		GoFiles           []string
+		CgoFiles          []string
+		TestGoFiles       []string
+		XTestGoFiles      []string
+		IgnoredGoFiles    []string
+		IgnoredOtherFiles []string
+	}
+
+	decoder := json.NewDecoder(stdout)
+	for decoder.More() {
+		var pkg packageInfo
+		if err := decoder.Decode(&pkg); err != nil {
+			return nil, fmt.Errorf("output malformed: %v", err)
+		}
+
+		for _, pkgFiles := range [][]string{
+			pkg.GoFiles,
+			pkg.CgoFiles,
+			pkg.TestGoFiles,
+			pkg.XTestGoFiles,
+			pkg.IgnoredGoFiles,
+			pkg.IgnoredOtherFiles,
+		} {
```

Comment:
@prashantv should these be filtered down to only `.go` files?

Reply:
Updated. Feel free to undo.

Reply:
Ah turns out `IgnoredOtherFiles` isn't needed -- dropped `IgnoredOtherFiles` and the filter